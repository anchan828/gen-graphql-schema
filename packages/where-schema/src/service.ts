/* eslint-disable @typescript-eslint/no-non-null-assertion */
import {
  appendDefinitionToDocumentNode,
  getDefinitionByName,
  getDirectives,
  getFieldDefinitions,
  getFieldDefinitionsByDirective,
  getFieldTypeName,
  getObjectOrUnionTypeDefinition,
  getObjectTypeDefinitionsFromUnion,
  hasDirectiveInDocumentNode,
  isBasicType,
  isEnumType,
  isUnionType,
  toConstanceCase,
} from "@anchan828/gen-graphql-schema-common";
import * as deepmerge from "deepmerge";
import {
  DirectiveNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  ObjectTypeDefinitionNode,
  parse,
  TypeNode,
  UnionTypeDefinitionNode,
} from "graphql";
import { DEFAULT_OPTIONS, DESCRIPTIONS } from "./constants";
import { GenWhereTypesOptions, WhereFieldNameAndType } from "./interfaces";
export class GenWhereTypesService {
  public genWhereTypes(): DocumentNode {
    if (!hasDirectiveInDocumentNode(this.documentNode, this.options.whereDirective!.name!)) {
      return this.documentNode;
    }
    const fields = getFieldDefinitionsByDirective(this.documentNode, this.options.whereDirective!.name!);
    for (const field of fields) {
      const whereType = this.genWhereTypeDefinition(field);

      if (!whereType) {
        continue;
      }

      this.appendWhereArgumentToFieldNode(field, whereType);
      this.removeWhereDirective(field);
      this.removeWhereIgnoreDirective(field);
    }

    for (const operator of this.typeOperatorMap.values()) {
      appendDefinitionToDocumentNode(this.documentNode, operator.enumType, operator.inputObjectType);
    }

    return this.documentNode;
  }

  private genOperatorDefinitions(typeName: string): void {
    if (this.typeOperatorMap.has(typeName)) {
      return;
    }
    let operatorNames = Reflect.get(this.options.supportOperatorTypes!, typeName) as string[];

    if (!Array.isArray(operatorNames) && isEnumType(this.documentNode, typeName)) {
      operatorNames = this.options.enumTypeOperator!;
    }
    if (!Array.isArray(operatorNames)) {
      return;
    }

    this.typeOperatorMap.set(typeName, {
      enumType: {
        kind: "EnumTypeDefinition",
        name: {
          kind: "Name",
          value: `${this.options.whereOperatorType!.prefix}${typeName}${this.options.whereOperatorType!.suffix}`,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.WHERE_OPERATOR_TYPE.TYPE(typeName),
        },
        values: operatorNames.map(
          operatorName =>
            ({
              kind: "EnumValueDefinition",
              name: {
                kind: "Name",
                value: toConstanceCase(operatorName),
              },
              description: {
                kind: "StringValue",
                value: Reflect.get(DESCRIPTIONS.WHERE_OPERATOR_TYPE.OPERATORS, toConstanceCase(operatorName)),
              },
            } as EnumValueDefinitionNode),
        ),
      },
      inputObjectType: {
        kind: "InputObjectTypeDefinition",
        name: {
          kind: "Name",
          value: `${this.options.whereOperator!.prefix}${typeName}${this.options.whereOperator!.suffix}`,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.WHERE_OPERATOR.TYPE(typeName),
        },
        fields: [
          {
            kind: "InputValueDefinition",
            name: {
              kind: "Name",
              value: "type",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.WHERE_OPERATOR.FIELDS.TYPE(typeName),
            },
            type: {
              kind: "NonNullType",
              type: {
                kind: "NamedType",
                name: {
                  kind: "Name",
                  value: `${this.options.whereOperatorType!.prefix}${typeName}${
                    this.options.whereOperatorType!.suffix
                  }`,
                },
              },
            },
          },
          {
            kind: "InputValueDefinition",
            name: {
              kind: "Name",
              value: "value",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.WHERE_OPERATOR.FIELDS.VALUE(typeName),
            },
            type: {
              kind: "ListType",
              type: {
                kind: "NamedType",
                name: {
                  kind: "Name",
                  value: typeName,
                },
              },
            },
          },
        ],
      } as InputObjectTypeDefinitionNode,
    });
  }

  private appendWhereArgumentToFieldNode(
    field: FieldDefinitionNode,
    whereTypeDefinition: InputObjectTypeDefinitionNode,
  ): void {
    Reflect.set(field, "arguments", [
      ...(field.arguments || []),
      {
        kind: "FieldDefinition",
        name: {
          kind: "Name",
          value: this.options.whereArgment!.name!,
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: whereTypeDefinition.name.value,
          },
        },
        directives: [],
      },
    ]);
  }

  private genWhereTypeDefinition(field: FieldDefinitionNode): InputObjectTypeDefinitionNode | undefined {
    const fieldType = getObjectOrUnionTypeDefinition(this.documentNode, field);
    if (!fieldType) {
      return;
    }

    const fieldNameAndTypes = this.getWhereFieldNameAndTypes(fieldType);
    const fields: InputValueDefinitionNode[] = [];
    for (const { name, type, isEqOnly, isObject } of fieldNameAndTypes) {
      let fieldTypeNode: TypeNode | undefined;

      if (type === "Boolean" || isEqOnly) {
        fieldTypeNode = {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: type,
          },
        };
      } else if (isObject) {
        this.genOperatorDefinitions(type);
        fieldTypeNode = {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: `${this.options.whereType!.prefix}${type}${this.options.whereType!.suffix}`,
          },
        };
      } else {
        this.genOperatorDefinitions(type);
        fieldTypeNode = {
          kind: "ListType",
          type: {
            kind: "NamedType",
            name: {
              kind: "Name",
              value: `${this.options.whereOperator!.prefix}${type}${this.options.whereOperator!.suffix}`,
            },
          },
        };
      }
      if (fieldTypeNode) {
        fields.push({
          kind: "InputValueDefinition",
          name: {
            kind: "Name",
            value: name,
          },
          description: {
            kind: "StringValue",
            value: DESCRIPTIONS.WHERE_TYPE.FIELDS(name),
          },
          type: fieldTypeNode,
          directives: [],
        });
      }
    }

    const whereTypeOptions = this.options.whereType!;
    const whereTypeName = `${whereTypeOptions.prefix}${getFieldTypeName(field).name}${whereTypeOptions.suffix}`;
    let whereType = getDefinitionByName(this.documentNode, whereTypeName) as InputObjectTypeDefinitionNode;
    if (whereType) {
      return whereType;
    }
    whereType = {
      kind: "InputObjectTypeDefinition",
      name: {
        kind: "Name",
        value: whereTypeName,
      },
      description: {
        kind: "StringValue",
        value: DESCRIPTIONS.WHERE_TYPE.TYPE(getFieldTypeName(field).name),
      },
      directives: [],
      fields,
    } as InputObjectTypeDefinitionNode;
    appendDefinitionToDocumentNode(this.documentNode, whereType);
    return whereType;
  }

  // TODO: ここを改善
  private getWhereFieldNameAndTypes(
    definition: ObjectTypeDefinitionNode | UnionTypeDefinitionNode,
  ): WhereFieldNameAndType[] {
    const fieldNames: WhereFieldNameAndType[] = [];

    const fields: FieldDefinitionNode[] = [];

    if (isUnionType(definition)) {
      for (const def of getObjectTypeDefinitionsFromUnion(this.documentNode, definition)) {
        fields.push(...getFieldDefinitions(def));
      }
    } else {
      fields.push(...getFieldDefinitions(definition));
    }

    for (const field of fields) {
      const { name, isList } = getFieldTypeName(field);

      if (fieldNames.findIndex(fn => fn.name === field.name.value) !== -1) {
        continue;
      }

      if (this.hasDirective(field, this.options.whereIgnoreDirective?.name)) {
        continue;
      }

      if (
        !(
          isBasicType(name) ||
          isEnumType(this.documentNode, name) ||
          Reflect.get(this.options.supportOperatorTypes!, name)
        )
      ) {
        if (name !== definition.name.value) {
          this.genWhereTypeDefinition(field);
        }
        fieldNames.push({
          name: field.name.value,
          type: name,
          isList,
          isObject: true,
          isEqOnly: this.hasDirective(field, this.options.whereEqOnlyDirective?.name),
        });
        continue;
      }

      fieldNames.push({
        name: field.name.value,
        type: name,
        isList,
        isObject: false,
        isEqOnly: this.hasDirective(field, this.options.whereEqOnlyDirective?.name),
      });
    }

    return fieldNames;
  }

  private hasDirective(field: FieldDefinitionNode, directiveName?: string): boolean {
    if (!directiveName) {
      return false;
    }

    const directives = getDirectives(field);
    if (directives.find((d: DirectiveNode) => d.name.value === directiveName)) {
      return true;
    }

    return false;
  }

  private removeDirectives(field: FieldDefinitionNode, directiveNames: (string | undefined)[]): void {
    const names: string[] = directiveNames.filter(
      (directiveName: string | undefined): directiveName is string => directiveName !== undefined,
    );
    Reflect.set(
      field,
      "directives",
      getDirectives(field).filter(directive => !names.includes(directive.name.value)),
    );
  }

  private removeWhereDirective(field: FieldDefinitionNode): void {
    const directives = getDirectives(field);
    Reflect.set(
      field,
      "directives",
      directives.filter(directive => directive.name.value !== this.options.whereDirective!.name),
    );
  }

  private removeWhereIgnoreDirective(field: FieldDefinitionNode): void {
    const type = getObjectOrUnionTypeDefinition(this.documentNode, field);
    if (!type) {
      return;
    }
    const definitions: ObjectTypeDefinitionNode[] = [];

    if (isUnionType(type)) {
      definitions.push(...getObjectTypeDefinitionsFromUnion(this.documentNode, type));
    } else {
      definitions.push(type);
    }
    for (const definition of definitions) {
      for (const f of getFieldDefinitions(definition)) {
        this.removeDirectives(f, [this.options.whereIgnoreDirective?.name, this.options.whereEqOnlyDirective?.name]);
      }
    }
  }

  private readonly documentNode: DocumentNode;

  private typeOperatorMap: Map<
    string,
    {
      enumType: EnumTypeDefinitionNode;
      inputObjectType: InputObjectTypeDefinitionNode;
    }
  > = new Map<
    string,
    {
      enumType: EnumTypeDefinitionNode;
      inputObjectType: InputObjectTypeDefinitionNode;
    }
  >();

  constructor(readonly types: string | DocumentNode, private readonly options: GenWhereTypesOptions = {}) {
    this.options = deepmerge(DEFAULT_OPTIONS, options);
    this.documentNode = typeof types === "string" ? parse(types) : types;
  }
}
