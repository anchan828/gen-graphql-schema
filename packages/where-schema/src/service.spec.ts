import { buildASTSchema, parse, printSchema } from "graphql";
import { GenWhereTypesService } from "./service";

describe("GenWhereTypesService", () => {
  it("should return same schema when no where directive", () => {
    const types = `type Test { id: ID }`;
    expect(printSchema(buildASTSchema(new GenWhereTypesService(types).genWhereTypes()))).toEqual(
      printSchema(buildASTSchema(parse(types))),
    );
  });

  it("should return added Where types when has where directive", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenWhereTypesService(
            [
              `scalar Date`,
              `enum TestEnum { A B }`,
              `type Test {
            id: ID!
            name: String!
            texts: [String]
            age: Int
            age2: Int @where_ignore
            created_At: Date
            range: Float
            published: Boolean
            test: TestEnum
            test2: [TestEnum]
        }`,
              `type Query { tests: [Test] @where}`,
            ].join(`\n`),
            {
              supportOperatorTypes: {
                Date: ["eq", "not_eq", "in", "not_in", "lt", "lte", "gt", "gte"],
              },
            },
          ).genWhereTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should return added Where types when has custom directive", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenWhereTypesService(
            parse(
              [
                `type Test {
            id: ID!
            name: String @foo
        }`,
                `type Query { tests: [Test] @fuga}`,
              ].join(`\n`),
            ),
            {
              whereArgument: {
                name: "Hoge",
              },
              whereDirective: {
                name: "fuga",
              },
              whereIgnoreDirective: {
                name: "foo",
              },
              whereOperator: {
                prefix: "PreOperator",
                suffix: "SufOperator",
              },
              whereOperatorType: {
                prefix: "PreOperatorType",
                suffix: "SufOperatorType",
              },
              whereType: {
                prefix: "PreWhereType",
                suffix: "SufWhereType",
              },
            },
          ).genWhereTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should return added Where types when has union type and where directive", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenWhereTypesService(
            [
              `type Test1 { id: ID, name: String! }`,
              `type Test2 { id: ID, age: Int! }`,
              `union Test = Test1 | Test2`,
              `type Query { tests: [Test] @where}`,
            ].join(`\n`),
            {
              supportOperatorTypes: {
                Date: ["eq", "not_eq", "in", "not_in", "lt", "lte", "gt", "gte"],
              },
            },
          ).genWhereTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should apply where_eq_only directive", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenWhereTypesService(
            [
              `type Test {
            id: ID!
            name: String! @where_eq_only
        }`,
              `type Query { tests: [Test] @where}`,
            ].join(`\n`),
            {
              supportOperatorTypes: {
                Date: ["eq", "not_eq", "in", "not_in", "lt", "lte", "gt", "gte"],
              },
            },
          ).genWhereTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });

  it("should deep where", () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenWhereTypesService(
            [
              `
type TestPosition {
  column: Int!
  line: Int!
  prev: TestPosition @where_nested
  next: TestPosition @where_nested
}
type TestSubPosition {
  column: Int!
  line: Int!
}         
type Test {
  id: ID!
  position: TestPosition! @where_nested
}`,
              `type Query { tests: [Test] @where}`,
            ].join(`\n`),
            {
              supportOperatorTypes: {
                Date: ["eq", "not_eq", "in", "not_in", "lt", "lte", "gt", "gte"],
              },
            },
          ).genWhereTypes(),
        ),
      ),
    ).toMatchSnapshot();
  });
});
