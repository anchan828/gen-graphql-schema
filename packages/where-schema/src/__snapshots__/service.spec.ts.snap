// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenWhereTypesService should apply where_eq_only directive 1`] = `
"type Test {
  id: ID!
  name: String!
}

type Query {
  tests(where: TestWhere): [Test]
}

\\"\\"\\"Query of Test with using operators\\"\\"\\"
input TestWhere {
  \\"\\"\\"Query with using id field\\"\\"\\"
  id: [IDWhereOperator]

  \\"\\"\\"Query with using name field\\"\\"\\"
  name: String
}

\\"\\"\\"Query type of ID with using operators\\"\\"\\"
enum IDWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"ID query with using operators\\"\\"\\"
input IDWhereOperator {
  \\"\\"\\"Query type of ID with using operators\\"\\"\\"
  type: IDWhereOperatorType!

  \\"\\"\\"Query value of ID\\"\\"\\"
  value: [ID]
}
"
`;

exports[`GenWhereTypesService should deep where 1`] = `
"type TestPosition {
  column: Int!
  line: Int!
  prev: TestPosition
  next: TestPosition
}

type TestSubPosition {
  column: Int!
  line: Int!
}

type Test {
  id: ID!
  position: TestPosition!
}

type Query {
  tests(where: TestWhere): [Test]
}

\\"\\"\\"Query of TestPosition with using operators\\"\\"\\"
input TestPositionWhere {
  \\"\\"\\"Query with using column field\\"\\"\\"
  column: [IntWhereOperator]

  \\"\\"\\"Query with using line field\\"\\"\\"
  line: [IntWhereOperator]

  \\"\\"\\"Query with using prev field\\"\\"\\"
  prev: TestPositionWhere

  \\"\\"\\"Query with using next field\\"\\"\\"
  next: TestPositionWhere
}

\\"\\"\\"Query of Test with using operators\\"\\"\\"
input TestWhere {
  \\"\\"\\"Query with using id field\\"\\"\\"
  id: [IDWhereOperator]

  \\"\\"\\"Query with using position field\\"\\"\\"
  position: TestPositionWhere
}

\\"\\"\\"Query type of Int with using operators\\"\\"\\"
enum IntWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN

  \\"\\"\\"Must be less than given value\\"\\"\\"
  LT

  \\"\\"\\"Must be less than or equal to given value\\"\\"\\"
  LTE

  \\"\\"\\"Must be greater than given value\\"\\"\\"
  GT

  \\"\\"\\"Must be greater than or equal to given value\\"\\"\\"
  GTE
}

\\"\\"\\"Int query with using operators\\"\\"\\"
input IntWhereOperator {
  \\"\\"\\"Query type of Int with using operators\\"\\"\\"
  type: IntWhereOperatorType!

  \\"\\"\\"Query value of Int\\"\\"\\"
  value: [Int]
}

\\"\\"\\"Query type of ID with using operators\\"\\"\\"
enum IDWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"ID query with using operators\\"\\"\\"
input IDWhereOperator {
  \\"\\"\\"Query type of ID with using operators\\"\\"\\"
  type: IDWhereOperatorType!

  \\"\\"\\"Query value of ID\\"\\"\\"
  value: [ID]
}
"
`;

exports[`GenWhereTypesService should return added Where types when has custom directive 1`] = `
"type Test {
  id: ID!
  name: String
}

type Query {
  tests(Hoge: PreWhereTypeTestSufWhereType): [Test]
}

\\"\\"\\"Query of Test with using operators\\"\\"\\"
input PreWhereTypeTestSufWhereType {
  \\"\\"\\"Query with using id field\\"\\"\\"
  id: [PreOperatorIDSufOperator]
}

\\"\\"\\"Query type of ID with using operators\\"\\"\\"
enum PreOperatorTypeIDSufOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"ID query with using operators\\"\\"\\"
input PreOperatorIDSufOperator {
  \\"\\"\\"Query type of ID with using operators\\"\\"\\"
  type: PreOperatorTypeIDSufOperatorType!

  \\"\\"\\"Query value of ID\\"\\"\\"
  value: [ID]
}
"
`;

exports[`GenWhereTypesService should return added Where types when has union type and where directive 1`] = `
"type Test1 {
  id: ID
  name: String!
}

type Test2 {
  id: ID
  age: Int!
}

union Test = Test1 | Test2

type Query {
  tests(where: TestWhere): [Test]
}

\\"\\"\\"Query of Test with using operators\\"\\"\\"
input TestWhere {
  \\"\\"\\"Query with using id field\\"\\"\\"
  id: [IDWhereOperator]

  \\"\\"\\"Query with using name field\\"\\"\\"
  name: [StringWhereOperator]

  \\"\\"\\"Query with using age field\\"\\"\\"
  age: [IntWhereOperator]
}

\\"\\"\\"Query type of ID with using operators\\"\\"\\"
enum IDWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"ID query with using operators\\"\\"\\"
input IDWhereOperator {
  \\"\\"\\"Query type of ID with using operators\\"\\"\\"
  type: IDWhereOperatorType!

  \\"\\"\\"Query value of ID\\"\\"\\"
  value: [ID]
}

\\"\\"\\"Query type of String with using operators\\"\\"\\"
enum StringWhereOperatorType {
  \\"\\"\\"Must match the string starts with the given data exactly\\"\\"\\"
  STARTS_WITH

  \\"\\"\\"Must match the string ends with the given data exactly\\"\\"\\"
  ENDS_WITH

  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"
  Determines whether the given string may be found within another string.
  \\"\\"\\"
  CONTAINS

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"String query with using operators\\"\\"\\"
input StringWhereOperator {
  \\"\\"\\"Query type of String with using operators\\"\\"\\"
  type: StringWhereOperatorType!

  \\"\\"\\"Query value of String\\"\\"\\"
  value: [String]
}

\\"\\"\\"Query type of Int with using operators\\"\\"\\"
enum IntWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN

  \\"\\"\\"Must be less than given value\\"\\"\\"
  LT

  \\"\\"\\"Must be less than or equal to given value\\"\\"\\"
  LTE

  \\"\\"\\"Must be greater than given value\\"\\"\\"
  GT

  \\"\\"\\"Must be greater than or equal to given value\\"\\"\\"
  GTE
}

\\"\\"\\"Int query with using operators\\"\\"\\"
input IntWhereOperator {
  \\"\\"\\"Query type of Int with using operators\\"\\"\\"
  type: IntWhereOperatorType!

  \\"\\"\\"Query value of Int\\"\\"\\"
  value: [Int]
}
"
`;

exports[`GenWhereTypesService should return added Where types when has where directive 1`] = `
"scalar Date

enum TestEnum {
  A
  B
}

type Test {
  id: ID!
  name: String!
  texts: [String]
  age: Int
  age2: Int
  created_At: Date
  range: Float
  published: Boolean
  test: TestEnum
  test2: [TestEnum]
}

type Query {
  tests(where: TestWhere): [Test]
}

\\"\\"\\"Query of Test with using operators\\"\\"\\"
input TestWhere {
  \\"\\"\\"Query with using id field\\"\\"\\"
  id: [IDWhereOperator]

  \\"\\"\\"Query with using name field\\"\\"\\"
  name: [StringWhereOperator]

  \\"\\"\\"Query with using texts field\\"\\"\\"
  texts: [StringWhereOperator]

  \\"\\"\\"Query with using age field\\"\\"\\"
  age: [IntWhereOperator]

  \\"\\"\\"Query with using created_At field\\"\\"\\"
  created_At: [DateWhereOperator]

  \\"\\"\\"Query with using range field\\"\\"\\"
  range: [FloatWhereOperator]

  \\"\\"\\"Query with using published field\\"\\"\\"
  published: Boolean

  \\"\\"\\"Query with using test field\\"\\"\\"
  test: [TestEnumWhereOperator]

  \\"\\"\\"Query with using test2 field\\"\\"\\"
  test2: [TestEnumWhereOperator]
}

\\"\\"\\"Query type of ID with using operators\\"\\"\\"
enum IDWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"ID query with using operators\\"\\"\\"
input IDWhereOperator {
  \\"\\"\\"Query type of ID with using operators\\"\\"\\"
  type: IDWhereOperatorType!

  \\"\\"\\"Query value of ID\\"\\"\\"
  value: [ID]
}

\\"\\"\\"Query type of String with using operators\\"\\"\\"
enum StringWhereOperatorType {
  \\"\\"\\"Must match the string starts with the given data exactly\\"\\"\\"
  STARTS_WITH

  \\"\\"\\"Must match the string ends with the given data exactly\\"\\"\\"
  ENDS_WITH

  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"
  Determines whether the given string may be found within another string.
  \\"\\"\\"
  CONTAINS

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"String query with using operators\\"\\"\\"
input StringWhereOperator {
  \\"\\"\\"Query type of String with using operators\\"\\"\\"
  type: StringWhereOperatorType!

  \\"\\"\\"Query value of String\\"\\"\\"
  value: [String]
}

\\"\\"\\"Query type of Int with using operators\\"\\"\\"
enum IntWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN

  \\"\\"\\"Must be less than given value\\"\\"\\"
  LT

  \\"\\"\\"Must be less than or equal to given value\\"\\"\\"
  LTE

  \\"\\"\\"Must be greater than given value\\"\\"\\"
  GT

  \\"\\"\\"Must be greater than or equal to given value\\"\\"\\"
  GTE
}

\\"\\"\\"Int query with using operators\\"\\"\\"
input IntWhereOperator {
  \\"\\"\\"Query type of Int with using operators\\"\\"\\"
  type: IntWhereOperatorType!

  \\"\\"\\"Query value of Int\\"\\"\\"
  value: [Int]
}

\\"\\"\\"Query type of Date with using operators\\"\\"\\"
enum DateWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN

  \\"\\"\\"Must be less than given value\\"\\"\\"
  LT

  \\"\\"\\"Must be less than or equal to given value\\"\\"\\"
  LTE

  \\"\\"\\"Must be greater than given value\\"\\"\\"
  GT

  \\"\\"\\"Must be greater than or equal to given value\\"\\"\\"
  GTE
}

\\"\\"\\"Date query with using operators\\"\\"\\"
input DateWhereOperator {
  \\"\\"\\"Query type of Date with using operators\\"\\"\\"
  type: DateWhereOperatorType!

  \\"\\"\\"Query value of Date\\"\\"\\"
  value: [Date]
}

\\"\\"\\"Query type of Float with using operators\\"\\"\\"
enum FloatWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN

  \\"\\"\\"Must be less than given value\\"\\"\\"
  LT

  \\"\\"\\"Must be less than or equal to given value\\"\\"\\"
  LTE

  \\"\\"\\"Must be greater than given value\\"\\"\\"
  GT

  \\"\\"\\"Must be greater than or equal to given value\\"\\"\\"
  GTE
}

\\"\\"\\"Float query with using operators\\"\\"\\"
input FloatWhereOperator {
  \\"\\"\\"Query type of Float with using operators\\"\\"\\"
  type: FloatWhereOperatorType!

  \\"\\"\\"Query value of Float\\"\\"\\"
  value: [Float]
}

\\"\\"\\"Query type of TestEnum with using operators\\"\\"\\"
enum TestEnumWhereOperatorType {
  \\"\\"\\"Must match the given data exactly\\"\\"\\"
  EQ

  \\"\\"\\"Must be different from the given data\\"\\"\\"
  NOT_EQ

  \\"\\"\\"Must be an element of the array\\"\\"\\"
  IN

  \\"\\"\\"Must not be an element of the array\\"\\"\\"
  NOT_IN
}

\\"\\"\\"TestEnum query with using operators\\"\\"\\"
input TestEnumWhereOperator {
  \\"\\"\\"Query type of TestEnum with using operators\\"\\"\\"
  type: TestEnumWhereOperatorType!

  \\"\\"\\"Query value of TestEnum\\"\\"\\"
  value: [TestEnum]
}
"
`;
